regression_em/model.py

```diff
@@ -1,18 +1,287 @@
+        """Estimate regression EM params.
+
+        :param X: {array-like, sparse matrix} of shape (n_samples, n_left and right latent features)
+                  Feature matrix derived from concatenating left latent features with right latent features.
+        :param y: array-like of shape (n_samples,)
+                  Sequence of labels indicating each sample is positive or negative.
+        """
+        # separate dataset with index.
+        if sp.issparse(X):
+            left_feat = X[:, 0:self.split_index]
+            right_feat = X[:, self.split_index:]
+        else:
+            left_feat, right_feat = np.hsplit(X, [self.split_index])
+
+        # convert y to boolean
+        labels = self.binary_to_bool(y)
+
+        # Initialize params (feature weight, intercept).
+        self.left_params = (np.random.rand(left_feat.shape[1]), random())
+        self.right_params = (np.random.rand(right_feat.shape[1]), random())
+
+        self.log_likelihoods = [self._calc_log_likelihood(left_feat, right_feat, labels)]
+        max_ll = self.log_likelihoods[-1]
+        best_left_params = self.left_params
+        best_right_params = self.right_params
+
+        sample_weights = None
+        if self.class_weights == 'balanced':
+            pos_ratio = np.count_nonzero(y) / y.size
+            sample_weights = np.array([1 / pos_ratio if l else 1 / (1 - pos_ratio) for l in labels])
+
+        for epoch in range(self.max_iter):
+            # Update left latent params
+            left_responsibilities = self.update_responsibilities(self.left_params, left_feat, self.right_params, right_feat, labels)
+            self.left_params = self.update_params(left_feat, left_responsibilities, sample_weights)
+
+            # Update right latent params
+            right_responsibilities = self.update_responsibilities(self.right_params, right_feat, self.left_params, left_feat, labels)
+            self.right_params = self.update_params(right_feat, right_responsibilities, sample_weights)
+
+            # calculating log likelihood and judging convergence
+            self.log_likelihoods.append(self._calc_log_likelihood(left_feat, right_feat, labels))
+
+            if max_ll < self.log_likelihoods[-1]:
+                max_ll = self.log_likelihoods[-1]
+                best_left_params = self.left_params
+                best_right_params = self.right_params
+
+        self.left_params = best_left_params
+        self.right_params = best_right_params
+
+    def predict_proba(self, X: Matrix) -> Matrix:
+        """Return predicted probabilities.
 
+        :param X: {array-like, sparse matrix} of shape (n_samples, n_left and right latent features)
+                  Feature matrix derived from concatenating left latent features with right latent features.
+        :return: Predicted probabilities.
+        """
+        # separate dataset with index
+        if sp.issparse(X):
+            left_feat = X[:, 0:self.split_index]
```

Comment:
`left_feat = X[:, :self.split_index]` is ok.

Reply:
Fixed it.